### PR TITLE
doc: improve documentation

### LINKED ISSUE #185: Improve documentation
Four critical areas:
- [ ] `iterators` example
- [ ] `prec_climber` example
- [ ] `PrecClimber` example (can be the same as above)
- [ ] any undocumented stuff

### LINKED ISSUE #297: Document generator

### LINKED ISSUE #298: Document meta

### LINKED ISSUE #299: Document vm

### LINKED ISSUE #335: Please add an example on how to define multiple `Parser` structs in the same crate
As in subject, I am struggling to understand how to define multiple `Parser` structs in the same crate, each one relying on its own grammar file.

### PR DESCRIPTION
**Pull request by tomtau**
added missing_docs warning etc. and documented the currently undocumented public APIs. added additional explanations in README, updated links, etc.

Closes #335
Closes #299
Closes #298
Closes #297
Closes #185

### CODE REVIEW

File: meta/src/validator.rs [hunk 1]
```diff
@@ -67,7 +70,12 @@ static BUILTINS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
     .collect::<HashSet<&str>>()
 });
 
-pub fn validate_pairs(pairs: Pairs<Rule>) -> Result<Vec<&str>, Vec<Error<Rule>>> {
+/// The parsed grammar fro common mistakes:
```
Comment (NoahTheDuke) [inline meta/src/validator.rs:73]: `fro`?

File: meta/src/validator.rs [hunk 2]
```diff
@@ -167,6 +178,7 @@ pub fn validate_already_defined(definitions: &Vec<Span>) -> Vec<Error<Rule>> {
     errors
 }
 
+/// Validates that the given `definitions` do not contain any undefined rules.å
```
Comment (NoahTheDuke) [inline meta/src/validator.rs:181]:
```suggestion
/// Validates that the given `definitions` do not contain any undefined rules.
```

File: meta/src/validator.rs [hunk 3]
```diff
@@ -191,6 +203,10 @@ pub fn validate_undefined<'i>(
     errors
 }
 
+/// It validates the abstract syntax tree for common mistakes:
```
Comment (NoahTheDuke) [inline meta/src/validator.rs:206]:
```suggestion
/// Validates the abstract syntax tree for common mistakes:
```

File: pest/src/lib.rs [hunk 4]
```diff
@@ -61,18 +68,258 @@
...
 
 extern crate alloc;
 #[cfg(feature = "std")]
 extern crate std;
-extern crate ucd_trie;
-
-#[cfg(feature = "pretty-print")]
-extern crate serde;
-#[cfg(feature = "pretty-print")]
-extern crate serde_json;
```
Comment (NoahTheDuke) [inline pest/src/lib.rs:75]: Is this intended?

Comment (tomtau) [reply pest/src/lib.rs:75]: Yes, as I put `#![warn(missing_docs, rust_2018_idioms, unused_qualifications)]`, it flagged some of these extern crate declarations that are redundant in post-2018 editions

**Pull request merged**